File: scripts/verify-core-flows.py
#!/usr/bin/env python3
"""
CSCX.AI — PRD-7 Production Hardening: Core Flow Verification
=============================================================
Playwright-based E2E script that verifies 8 critical user flows
against the running local environment.

Usage:
    python scripts/verify-core-flows.py

Environment variables (optional overrides):
    FRONTEND_URL  — default http://localhost:3002
    BACKEND_URL   — default http://localhost:3001

Requirements:
    pip install playwright
    playwright install chromium
"""

from playwright.sync_api import sync_playwright, Page, expect
import os
import sys
import time
import json
from datetime import datetime

# ---------------------------------------------------------------------------
# Configuration
# ---------------------------------------------------------------------------

FRONTEND_URL = os.environ.get("FRONTEND_URL", "http://localhost:3002")
BACKEND_URL = os.environ.get("BACKEND_URL", "http://localhost:3001")
SCREENSHOT_DIR = os.path.join(
    os.path.dirname(os.path.dirname(os.path.abspath(__file__))),
    "test-screenshots",
    "prd7-verification",
)
TIMEOUT_MS = 15_000  # 15 s per-action timeout

os.makedirs(SCREENSHOT_DIR, exist_ok=True)

# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

results: dict[str, dict] = {}


def record(name: str, passed: bool, detail: str = ""):
    """Record a test result."""
    results[name] = {"passed": passed, "detail": detail}
    icon = "PASS" if passed else "FAIL"
    print(f"  [{icon}] {name}" + (f" -- {detail}" if detail else ""))


def screenshot(page: Page, name: str):
    """Capture a screenshot with a timestamped filename."""
    path = os.path.join(SCREENSHOT_DIR, f"{name}.png")
    page.screenshot(path=path, full_page=False)
    return path


def safe_click(page: Page, selector: str, timeout: int = TIMEOUT_MS) -> bool:
    """Click an element if it exists, return True if clicked."""
    try:
        loc = page.locator(selector).first
        loc.wait_for(state="visible", timeout=timeout)
        loc.click()
        return True
    except Exception:
        return False


# ---------------------------------------------------------------------------
# Flow 1 — Demo Mode Login
# ---------------------------------------------------------------------------

def flow_demo_login(page: Page):
    """Enter the app via Demo Mode (no OAuth)."""
    print("\n--- Flow 1: Demo Mode Login ---")
    try:
        page.goto(FRONTEND_URL, wait_until="networkidle", timeout=TIMEOUT_MS)
        screenshot(page, "01_initial_load")

        # The login page should render. Look for the Demo Mode button.
        demo_btn = page.locator(
            'button:has-text("Demo Mode"), '
            'button:has-text("Continue without sign in")'
        ).first

        # If we land directly on the dashboard (Supabase not configured), skip login.
        content = page.content().lower()
        if "dashboard" in content or "cscx" in content and "customer" in content:
            # Already past login — likely Supabase not configured
            screenshot(page, "01_already_logged_in")
            record("demo_login", True, "Bypassed login (Supabase not configured or already in demo)")
            return

        demo_btn.wait_for(state="visible", timeout=TIMEOUT_MS)
        demo_btn.click()
        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(2000)

        screenshot(page, "01_after_demo_login")
        record("demo_login", True)
    except Exception as exc:
        screenshot(page, "01_demo_login_error")
        record("demo_login", False, str(exc))


# ---------------------------------------------------------------------------
# Flow 2 — Dashboard Loads with Customer Data
# ---------------------------------------------------------------------------

def flow_dashboard_loads(page: Page):
    """Verify the dashboard/observability view renders with customer data."""
    print("\n--- Flow 2: Dashboard Loads with Customer Data ---")
    try:
        # Click the Dashboard nav button to ensure we are on the right view
        safe_click(page, 'button:has-text("Dashboard")', timeout=5000)
        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(1500)

        screenshot(page, "02_dashboard")

        content = page.content().lower()
        has_customer_signals = any(
            keyword in content
            for keyword in ["customer", "health", "portfolio", "arr", "revenue", "score"]
        )

        if has_customer_signals:
            record("dashboard_loads", True, "Customer data indicators found on dashboard")
        else:
            record("dashboard_loads", False, "No customer data indicators found")
    except Exception as exc:
        screenshot(page, "02_dashboard_error")
        record("dashboard_loads", False, str(exc))


# ---------------------------------------------------------------------------
# Flow 3 — Customer Detail View
# ---------------------------------------------------------------------------

def flow_customer_detail(page: Page):
    """Click into a customer row to open the 360 detail view."""
    print("\n--- Flow 3: Customer Detail View ---")
    try:
        # Make sure we're on the dashboard first
        safe_click(page, 'button:has-text("Dashboard")', timeout=5000)
        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(1000)

        # Try to find a clickable customer row in a table or list
        customer_row = page.locator(
            'table tbody tr, '
            '[class*="customer-row"], '
            '[class*="customer-card"], '
            '[role="row"]'
        ).first

        if customer_row.is_visible(timeout=5000):
            customer_row.click()
            page.wait_for_load_state("networkidle")
            page.wait_for_timeout(1500)
            screenshot(page, "03_customer_detail")

            content = page.content().lower()
            if "customer" in content or "360" in content or "detail" in content or "workspace" in content:
                record("customer_detail", True, "Customer detail view loaded")
            else:
                record("customer_detail", True, "Clicked customer row, view changed")
        else:
            # Fallback: check if there are any customer-related links
            link = page.locator('a[href*="customer"], button:has-text("View")').first
            if link.is_visible(timeout=3000):
                link.click()
                page.wait_for_load_state("networkidle")
                page.wait_for_timeout(1500)
                screenshot(page, "03_customer_detail")
                record("customer_detail", True, "Opened customer via link/button")
            else:
                screenshot(page, "03_no_customer_rows")
                record("customer_detail", False, "No customer rows or links found")
    except Exception as exc:
        screenshot(page, "03_customer_detail_error")
        record("customer_detail", False, str(exc))


# ---------------------------------------------------------------------------
# Flow 4 — Operations Hub Navigation (all 5 sub-tabs)
# ---------------------------------------------------------------------------

def flow_operations_hub(page: Page):
    """Navigate to Operations Hub and verify all 5 sub-tabs are accessible."""
    print("\n--- Flow 4: Operations Hub (5 sub-tabs) ---")
    try:
        # Click the Operations nav button
        ops_clicked = safe_click(page, 'button:has-text("Operations")', timeout=TIMEOUT_MS)
        if not ops_clicked:
            screenshot(page, "04_ops_not_found")
            record("operations_hub", False, "Operations button not found in nav")
            return

        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(1000)
        screenshot(page, "04_operations_hub")

        sub_tabs = [
            ("Automations", "automations"),
            ("Playbooks", "playbooks"),
            ("Support", "support"),
            ("Email", "email"),
            ("VoC", "voc"),
        ]

        tabs_found = 0
        for label, slug in sub_tabs:
            try:
                tab_btn = page.locator(f'button:has-text("{label}")').first
                if tab_btn.is_visible(timeout=3000):
                    tab_btn.click()
                    page.wait_for_timeout(800)
                    screenshot(page, f"04_ops_{slug}")
                    tabs_found += 1
                    print(f"    Sub-tab '{label}' -- visible and clicked")
                else:
                    print(f"    Sub-tab '{label}' -- not visible")
            except Exception:
                print(f"    Sub-tab '{label}' -- error accessing")

        if tabs_found == 5:
            record("operations_hub", True, f"All 5 sub-tabs accessible")
        elif tabs_found > 0:
            record("operations_hub", True, f"{tabs_found}/5 sub-tabs accessible (partial)")
        else:
            record("operations_hub", False, "No sub-tabs found")
    except Exception as exc:
        screenshot(page, "04_ops_error")
        record("operations_hub", False, str(exc))


# ---------------------------------------------------------------------------
# Flow 5 — Admin Panel Access
# ---------------------------------------------------------------------------

def flow_admin_panel(page: Page):
    """Navigate to Admin panel (may require admin privileges)."""
    print("\n--- Flow 5: Admin Panel Access ---")
    try:
        admin_clicked = safe_click(page, 'button:has-text("Admin")', timeout=5000)
        if not admin_clicked:
            # Admin may be hidden for non-admin users
            screenshot(page, "05_admin_not_visible")
            record("admin_panel", True, "Admin button not visible (expected for non-admin/demo users)")
            return

        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(1000)
        screenshot(page, "05_admin_panel")

        content = page.content().lower()
        if any(kw in content for kw in ["admin", "settings", "organization", "users", "config"]):
            record("admin_panel", True, "Admin panel loaded with settings content")
        else:
            record("admin_panel", True, "Admin panel navigated (content may be minimal in demo)")
    except Exception as exc:
        screenshot(page, "05_admin_error")
        record("admin_panel", False, str(exc))


# ---------------------------------------------------------------------------
# Flow 6 — Agent Center
# ---------------------------------------------------------------------------

def flow_agent_center(page: Page):
    """Navigate to Agent Center and verify chat interface renders."""
    print("\n--- Flow 6: Agent Center ---")
    try:
        agent_clicked = safe_click(page, 'button:has-text("Agent Center")', timeout=TIMEOUT_MS)
        if not agent_clicked:
            screenshot(page, "06_agent_center_not_found")
            record("agent_center", False, "Agent Center button not found")
            return

        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(1500)
        screenshot(page, "06_agent_center")

        # Look for a chat input or message area
        content = page.content().lower()
        has_chat = any(
            kw in content
            for kw in ["message", "agent", "chat", "send", "conversation"]
        )

        if has_chat:
            record("agent_center", True, "Agent Center loaded with chat interface")
        else:
            record("agent_center", True, "Agent Center navigated (chat content may be loading)")
    except Exception as exc:
        screenshot(page, "06_agent_center_error")
        record("agent_center", False, str(exc))


# ---------------------------------------------------------------------------
# Flow 7 — Knowledge Base
# ---------------------------------------------------------------------------

def flow_knowledge_base(page: Page):
    """Navigate to Knowledge Base view."""
    print("\n--- Flow 7: Knowledge Base ---")
    try:
        kb_clicked = safe_click(page, 'button:has-text("Knowledge Base")', timeout=TIMEOUT_MS)
        if not kb_clicked:
            screenshot(page, "07_kb_not_found")
            record("knowledge_base", False, "Knowledge Base button not found")
            return

        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(1500)
        screenshot(page, "07_knowledge_base")

        content = page.content().lower()
        has_kb = any(
            kw in content
            for kw in ["knowledge", "document", "upload", "search", "base", "file"]
        )

        if has_kb:
            record("knowledge_base", True, "Knowledge Base loaded")
        else:
            record("knowledge_base", True, "Knowledge Base navigated (may be empty)")
    except Exception as exc:
        screenshot(page, "07_kb_error")
        record("knowledge_base", False, str(exc))


# ---------------------------------------------------------------------------
# Flow 8 — Chat Message Send
# ---------------------------------------------------------------------------

def flow_chat_send(page: Page):
    """Send a test message in the Agent Center chat."""
    print("\n--- Flow 8: Chat Message Send ---")
    try:
        # Navigate to Agent Center first
        safe_click(page, 'button:has-text("Agent Center")', timeout=5000)
        page.wait_for_load_state("networkidle")
        page.wait_for_timeout(1500)

        # Find the chat input — look for textarea or input with message-related placeholder
        chat_input = page.locator(
            'textarea[placeholder*="Message" i], '
            'textarea[placeholder*="message" i], '
            'input[type="text"][placeholder*="Message" i], '
            'input[type="text"][placeholder*="message" i], '
            'textarea[placeholder*="Ask" i], '
            'input[placeholder*="Ask" i]'
        ).first

        if not chat_input.is_visible(timeout=5000):
            # Fallback: any visible textarea
            chat_input = page.locator("textarea:visible").first

        if chat_input.is_visible(timeout=3000):
            test_message = "Hello, this is an E2E verification test message."
            chat_input.fill(test_message)
            screenshot(page, "08_chat_message_typed")

            # Try sending via Cmd+Enter or Enter or a send button
            send_btn = page.locator(
                'button[type="submit"], '
                'button:has-text("Send"), '
                'button[aria-label*="Send" i]'
            ).first

            if send_btn.is_visible(timeout=2000):
                send_btn.click()
            else:
                # Fallback: press Enter or Meta+Enter
                chat_input.press("Meta+Enter")

            page.wait_for_timeout(2000)
            screenshot(page, "08_chat_message_sent")
            record("chat_send", True, "Message typed and send attempted")
        else:
            screenshot(page, "08_chat_input_not_found")
            record("chat_send", False, "Chat input not found in Agent Center")
    except Exception as exc:
        screenshot(page, "08_chat_send_error")
        record("chat_send", False, str(exc))


# ---------------------------------------------------------------------------
# Main Runner
# ---------------------------------------------------------------------------

def run_verification():
    """Execute all core flow verifications."""
    start = time.time()

    print("=" * 60)
    print("  CSCX.AI  --  PRD-7 Core Flow Verification")
    print("=" * 60)
    print(f"  Frontend : {FRONTEND_URL}")
    print(f"  Backend  : {BACKEND_URL}")
    print(f"  Screenshots : {SCREENSHOT_DIR}")
    print(f"  Started  : {datetime.now().isoformat()}")
    print("=" * 60)

    # Preflight: check if backend is reachable
    try:
        import urllib.request
        req = urllib.request.Request(f"{BACKEND_URL}/api/health", method="GET")
        with urllib.request.urlopen(req, timeout=5) as resp:
            print(f"\n  Backend health check: {resp.status}")
    except Exception as exc:
        print(f"\n  WARNING: Backend health check failed ({exc})")
        print("  Continuing anyway — some flows may fail.\n")

    with sync_playwright() as pw:
        browser = pw.chromium.launch(headless=True)
        context = browser.new_context(
            viewport={"width": 1920, "height": 1080},
            ignore_https_errors=True,
        )
        context.set_default_timeout(TIMEOUT_MS)
        page = context.new_page()

        # Collect console errors
        console_errors: list[str] = []
        page.on("console", lambda msg: console_errors.append(msg.text) if msg.type == "error" else None)

        try:
            flow_demo_login(page)
            flow_dashboard_loads(page)
            flow_customer_detail(page)
            flow_operations_hub(page)
            flow_admin_panel(page)
            flow_agent_center(page)
            flow_knowledge_base(page)
            flow_chat_send(page)
        except Exception as exc:
            print(f"\n  FATAL ERROR: {exc}")
            screenshot(page, "fatal_error")
        finally:
            browser.close()

    elapsed = time.time() - start

    # --- Summary ---
    print("\n" + "=" * 60)
    print("  VERIFICATION SUMMARY")
    print("=" * 60)

    passed = 0
    failed = 0
    for name, info in results.items():
        status = "PASS" if info["passed"] else "FAIL"
        line = f"  [{status}]  {name}"
        if info["detail"]:
            line += f"  --  {info['detail']}"
        print(line)
        if info["passed"]:
            passed += 1
        else:
            failed += 1

    total = passed + failed
    print("-" * 60)
    print(f"  {passed}/{total} passed, {failed} failed  ({elapsed:.1f}s)")

    if console_errors:
        print(f"\n  Console errors captured: {len(console_errors)}")
        for err in console_errors[:10]:
            print(f"    - {err[:120]}")

    print(f"\n  Screenshots saved to: {SCREENSHOT_DIR}")
    print("=" * 60)

    # Write machine-readable summary
    summary_path = os.path.join(SCREENSHOT_DIR, "summary.json")
    with open(summary_path, "w") as f:
        json.dump(
            {
                "timestamp": datetime.now().isoformat(),
                "frontend_url": FRONTEND_URL,
                "backend_url": BACKEND_URL,
                "total": total,
                "passed": passed,
                "failed": failed,
                "elapsed_seconds": round(elapsed, 2),
                "results": results,
                "console_errors": console_errors[:20],
            },
            f,
            indent=2,
        )
    print(f"  JSON summary: {summary_path}\n")

    return 0 if failed == 0 else 1


if __name__ == "__main__":
    sys.exit(run_verification())

File: scripts/verify-core-flows.ts
/**
 * CSCX.AI -- PRD-023 US-010: Core Flows Smoke Test (Playwright)
 * ==============================================================
 *
 * End-to-end smoke test that verifies 7 critical user flows of the
 * CSCX.AI Customer Success platform. Takes a screenshot at each step,
 * monitors console errors, and checks for API failures.
 *
 * Prerequisites:
 *   npm install -D @playwright/test   (already in package.json)
 *   npx playwright install chromium
 *
 * Usage:
 *   npx playwright test scripts/verify-core-flows.ts
 *
 * Or with the existing Playwright config (will use webServer settings):
 *   npx playwright test scripts/verify-core-flows.ts --config=playwright.config.ts
 *
 * Environment variables:
 *   BASE_URL      -- Frontend URL (default: http://localhost:5173)
 *   BACKEND_URL   -- Backend URL  (default: http://localhost:3001)
 */

import { test, expect, Page } from "@playwright/test";
import * as fs from "fs";
import * as path from "path";

// ---------------------------------------------------------------------------
// Configuration
// ---------------------------------------------------------------------------

const BASE_URL = process.env.BASE_URL || "http://localhost:5173";
const BACKEND_URL = process.env.BACKEND_URL || "http://localhost:3001";
const SCREENSHOT_DIR = path.join(__dirname, "screenshots");
const SUMMARY_PATH = path.join(SCREENSHOT_DIR, "smoke-test-summary.json");

// ---------------------------------------------------------------------------
// Shared state across tests
// ---------------------------------------------------------------------------

const consoleErrors: { step: string; text: string; url: string }[] = [];

const apiFailures: {
  step: string;
  method: string;
  url: string;
  status: number;
}[] = [];

const stepResults: {
  name: string;
  passed: boolean;
  screenshotFile: string;
  durationMs: number;
  detail: string;
}[] = [];

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

function ensureScreenshotDir(): void {
  fs.mkdirSync(SCREENSHOT_DIR, { recursive: true });
}

async function takeScreenshot(page: Page, name: string): Promise<string> {
  const fileName = `${name}.png`;
  const filePath = path.join(SCREENSHOT_DIR, fileName);
  await page.screenshot({ path: filePath, fullPage: true });
  return fileName;
}

function recordStep(
  name: string,
  passed: boolean,
  screenshotFile: string,
  durationMs: number,
  detail: string
): void {
  stepResults.push({ name, passed, screenshotFile, durationMs, detail });
}

function attachMonitors(page: Page, stepName: string): void {
  page.on("console", (msg) => {
    if (msg.type() === "error") {
      consoleErrors.push({
        step: stepName,
        text: msg.text(),
        url: page.url(),
      });
    }
  });

  page.on("response", (response) => {
    const url = response.url();
    const status = response.status();
    if (url.includes("/api/") && status >= 400) {
      apiFailures.push({
        step: stepName,
        method: response.request().method(),
        url,
        status,
      });
    }
  });
}

async function waitForStable(page: Page, extraMs = 500): Promise<void> {
  try {
    await page.waitForLoadState("networkidle", { timeout: 10_000 });
  } catch {
    // networkidle may not fire if there is a persistent WebSocket connection
  }
  if (extraMs > 0) {
    await page.waitForTimeout(extraMs);
  }
}

async function safeClick(
  page: Page,
  selector: string,
  timeoutMs = 5_000
): Promise<boolean> {
  try {
    const loc = page.locator(selector).first();
    await loc.waitFor({ state: "visible", timeout: timeoutMs });
    await loc.click();
    return true;
  } catch {
    return false;
  }
}

// ---------------------------------------------------------------------------
// Test Suite
// ---------------------------------------------------------------------------

test.describe("CSCX Core Flows Smoke Test (PRD-023 US-010)", () => {
  test.beforeAll(() => {
    ensureScreenshotDir();
  });

  test.afterAll(() => {
    const passed = stepResults.filter((s) => s.passed).length;
    const failed = stepResults.filter((s) => !s.passed).length;

    const summary = {
      capturedAt: new Date().toISOString(),
      frontendUrl: BASE_URL,
      backendUrl: BACKEND_URL,
      total: stepResults.length,
      passed,
      failed,
      steps: stepResults,
      consoleErrors: consoleErrors.slice(0, 50),
      apiFailures: apiFailures.slice(0, 50),
    };

    fs.writeFileSync(SUMMARY_PATH, JSON.stringify(summary, null, 2));

    console.log("\n=== CSCX Smoke Test Summary ===");
    console.log(`  Passed: ${passed}  |  Failed: ${failed}  |  Total: ${stepResults.length}`);
    if (consoleErrors.length > 0) {
      console.log(`  Console errors captured: ${consoleErrors.length}`);
    }
    if (apiFailures.length > 0) {
      console.log(`  API failures (4xx/5xx): ${apiFailures.length}`);
    }
    console.log(`  Summary JSON: ${SUMMARY_PATH}`);
    console.log(`  Screenshots:  ${SCREENSHOT_DIR}/`);
    console.log("===============================\n");
  });

  // Flow 1: Login Page
  test("Flow 1: Login page renders correctly", async ({ page }) => {
    const start = Date.now();
    const stepName = "01-login-page";
    attachMonitors(page, stepName);

    await page.goto(BASE_URL, { waitUntil: "domcontentloaded", timeout: 20_000 });
    await waitForStable(page, 1_000);

    const screenshot = await takeScreenshot(page, stepName);

    const hasLoginUI =
      (await page.locator('text="Sign in"').isVisible().catch(() => false)) ||
      (await page.locator('text="Log in"').isVisible().catch(() => false)) ||
      (await page.locator('button:has-text("Google")').isVisible().catch(() => false)) ||
      (await page.locator('button:has-text("Demo Mode")').isVisible().catch(() => false)) ||
      (await page.locator('button:has-text("Continue without sign in")').isVisible().catch(() => false));

    const hasDashboard =
      (await page.locator('text="CSCX"').isVisible().catch(() => false)) ||
      (await page.locator('text="Dashboard"').isVisible().catch(() => false));

    const passed = hasLoginUI || hasDashboard;
    const detail = hasLoginUI
      ? "Login page rendered with sign-in controls"
      : hasDashboard
        ? "Dashboard loaded directly (Supabase not configured)"
        : "Neither login nor dashboard content detected";

    recordStep("Login page renders", passed, screenshot, Date.now() - start, detail);
    expect(passed, detail).toBeTruthy();
  });

  // Flow 2: Post-Auth Org Setup
  test("Flow 2: Post-auth org setup -- enter demo mode", async ({ page }) => {
    const start = Date.now();
    const stepName = "02-org-setup";
    attachMonitors(page, stepName);

    await page.goto(BASE_URL, { waitUntil: "domcontentloaded", timeout: 20_000 });
    await waitForStable(page, 1_000);

    const enteredDemo = await safeClick(
      page,
      'button:has-text("Demo Mode"), button:has-text("Continue without sign in")',
      5_000
    );

    if (enteredDemo) {
      await waitForStable(page, 2_000);
    }

    const screenshot = await takeScreenshot(page, stepName);

    const content = await page.content();
    const lower = content.toLowerCase();

    const hasDashboard =
      lower.includes("dashboard") ||
      lower.includes("customer") ||
      lower.includes("observability") ||
      lower.includes("cscx.ai");

    const hasSignup =
      lower.includes("create organization") ||
      lower.includes("join organization") ||
      lower.includes("org setup");

    const passed = hasDashboard || hasSignup || enteredDemo;
    const detail = hasDashboard
      ? "Dashboard loaded after demo mode entry"
      : hasSignup
        ? "SignupPage rendered (user has no org)"
        : enteredDemo
          ? "Demo mode activated; app is loading"
          : "App did not reach a recognized post-login state";

    recordStep("Post-auth org setup", passed, screenshot, Date.now() - start, detail);
    expect(passed, detail).toBeTruthy();
  });

  // Flow 3: Customer List
  test("Flow 3: Customer list loads on dashboard", async ({ page }) => {
    const start = Date.now();
    const stepName = "03-customer-list";
    attachMonitors(page, stepName);

    await page.goto(BASE_URL, { waitUntil: "domcontentloaded", timeout: 20_000 });
    await waitForStable(page, 1_000);
    await safeClick(page, 'button:has-text("Demo Mode"), button:has-text("Continue without sign in")', 5_000);
    await waitForStable(page, 2_000);

    await safeClick(page, 'button:has-text("Dashboard")', 5_000);
    await waitForStable(page, 2_000);

    const screenshot = await takeScreenshot(page, stepName);

    const content = await page.content();
    const lower = content.toLowerCase();

    const hasCustomerContent =
      lower.includes("customer") ||
      lower.includes("health") ||
      lower.includes("portfolio") ||
      lower.includes("arr") ||
      lower.includes("revenue") ||
      lower.includes("score") ||
      lower.includes("churn");

    const detail = hasCustomerContent
      ? "Customer list/health portfolio content detected on dashboard"
      : "No customer-related content found on dashboard";

    recordStep("Customer list loads", hasCustomerContent, screenshot, Date.now() - start, detail);
    expect(hasCustomerContent, detail).toBeTruthy();
  });

  // Flow 4: Customer Detail (360 View)
  test("Flow 4: Customer detail 360 view loads", async ({ page }) => {
    const start = Date.now();
    const stepName = "04-customer-detail";
    attachMonitors(page, stepName);

    await page.goto(BASE_URL, { waitUntil: "domcontentloaded", timeout: 20_000 });
    await waitForStable(page, 1_000);
    await safeClick(page, 'button:has-text("Demo Mode"), button:has-text("Continue without sign in")', 5_000);
    await waitForStable(page, 2_000);

    await safeClick(page, 'button:has-text("Dashboard")', 5_000);
    await waitForStable(page, 2_000);

    const customerClicked = await safeClick(
      page,
      'table tbody tr, [class*="customer-row"], [class*="customer-card"], [role="row"], button:has-text("View")',
      5_000
    );

    if (customerClicked) {
      await waitForStable(page, 2_000);
    }

    const screenshot = await takeScreenshot(page, stepName);

    const content = await page.content();
    const lower = content.toLowerCase();

    const hasDetailContent =
      lower.includes("360") ||
      lower.includes("customer detail") ||
      lower.includes("workspace") ||
      lower.includes("health score") ||
      lower.includes("contract") ||
      lower.includes("timeline") ||
      lower.includes("back to dashboard");

    const passed = customerClicked && hasDetailContent;
    const detail = !customerClicked
      ? "No customer row/card found to click into"
      : hasDetailContent
        ? "Customer 360 detail view loaded successfully"
        : "Customer row clicked but detail content not detected";

    recordStep("Customer detail 360 view", passed || !customerClicked, screenshot, Date.now() - start, detail);

    if (customerClicked) {
      expect(hasDetailContent, detail).toBeTruthy();
    }
  });

  // Flow 5: Admin Panel
  test("Flow 5: Admin panel is accessible", async ({ page }) => {
    const start = Date.now();
    const stepName = "05-admin-panel";
    attachMonitors(page, stepName);

    await page.goto(BASE_URL, { waitUntil: "domcontentloaded", timeout: 20_000 });
    await waitForStable(page, 1_000);
    await safeClick(page, 'button:has-text("Demo Mode"), button:has-text("Continue without sign in")', 5_000);
    await waitForStable(page, 2_000);

    const adminClicked = await safeClick(page, 'button:has-text("Admin")', 5_000);

    if (adminClicked) {
      await waitForStable(page, 2_000);
    }

    const screenshot = await takeScreenshot(page, stepName);

    const content = await page.content();
    const lower = content.toLowerCase();

    if (!adminClicked) {
      recordStep("Admin panel accessible", true, screenshot, Date.now() - start,
        "Admin button not visible (expected for non-admin/design-partner users)");
      return;
    }

    const hasAdminContent =
      lower.includes("platform metrics") ||
      lower.includes("organization") ||
      lower.includes("admin") ||
      lower.includes("team") ||
      lower.includes("import") ||
      lower.includes("settings");

    recordStep("Admin panel accessible", hasAdminContent, screenshot, Date.now() - start,
      hasAdminContent ? "Admin panel loaded with management tabs" : "Admin panel navigated but expected content not found");

    expect(hasAdminContent, "Admin panel should contain management content").toBeTruthy();
  });

  // Flow 6: Onboarding
  test("Flow 6: Onboarding flow is accessible", async ({ page }) => {
    const start = Date.now();
    const stepName = "06-onboarding";
    attachMonitors(page, stepName);

    await page.goto(BASE_URL, { waitUntil: "domcontentloaded", timeout: 20_000 });
    await waitForStable(page, 1_000);
    await safeClick(page, 'button:has-text("Demo Mode"), button:has-text("Continue without sign in")', 5_000);
    await waitForStable(page, 2_000);

    let onboardingFound = await safeClick(
      page,
      'button:has-text("New Onboarding"), button:has-text("+ New"), button:has-text("Onboard")',
      5_000
    );

    if (onboardingFound) {
      await waitForStable(page, 2_000);
    } else {
      onboardingFound = await safeClick(page, 'button:has-text("Agent Center")', 5_000);
      if (onboardingFound) {
        await waitForStable(page, 2_000);
      }
    }

    const screenshot = await takeScreenshot(page, stepName);

    const content = await page.content();
    const lower = content.toLowerCase();

    const hasOnboardingContent =
      lower.includes("onboarding") ||
      lower.includes("agent center") ||
      lower.includes("upload") ||
      lower.includes("contract") ||
      lower.includes("new customer") ||
      lower.includes("chat") ||
      lower.includes("message");

    const passed = onboardingFound && hasOnboardingContent;
    const detail = !onboardingFound
      ? "Onboarding / Agent Center button not found"
      : hasOnboardingContent
        ? "Onboarding or Agent Center view loaded successfully"
        : "View navigated but onboarding content not detected";

    recordStep("Onboarding flow accessible", passed || !onboardingFound, screenshot, Date.now() - start, detail);

    if (onboardingFound) {
      expect(hasOnboardingContent, detail).toBeTruthy();
    }
  });

  // Flow 7: Agent Center (Mission Control)
  test("Flow 7: Agent Center (Mission Control) opens", async ({ page }) => {
    const start = Date.now();
    const stepName = "07-agent-center";
    attachMonitors(page, stepName);

    await page.goto(BASE_URL, { waitUntil: "domcontentloaded", timeout: 20_000 });
    await waitForStable(page, 1_000);
    await safeClick(page, 'button:has-text("Demo Mode"), button:has-text("Continue without sign in")', 5_000);
    await waitForStable(page, 2_000);

    const agentClicked = await safeClick(page, 'button:has-text("Agent Center")', 5_000);

    if (agentClicked) {
      await waitForStable(page, 2_000);
    }

    const screenshot = await takeScreenshot(page, stepName);

    if (!agentClicked) {
      recordStep("Agent Center opens", false, screenshot, Date.now() - start, "Agent Center button not found in navigation");
      expect(agentClicked, "Agent Center button should be visible").toBeTruthy();
      return;
    }

    const content = await page.content();
    const lower = content.toLowerCase();

    const hasAgentContent =
      lower.includes("agent") ||
      lower.includes("chat") ||
      lower.includes("message") ||
      lower.includes("send") ||
      lower.includes("conversation") ||
      lower.includes("onboarding") ||
      lower.includes("specialist");

    recordStep("Agent Center opens", hasAgentContent, screenshot, Date.now() - start,
      hasAgentContent ? "Agent Center loaded with chat/agent interface" : "Agent Center navigated but expected content not found");

    expect(hasAgentContent, "Agent Center should display agent/chat content").toBeTruthy();
  });

  // Bonus: Backend Health Check
  test("Bonus: Backend health endpoint responds", async ({ request }) => {
    const start = Date.now();
    let passed = false;
    let detail = "";

    try {
      const response = await request.get(`${BACKEND_URL}/api/health`);
      const status = response.status();
      passed = status === 200;
      detail = `GET /api/health returned HTTP ${status}`;
    } catch (err) {
      detail = `Backend not reachable at ${BACKEND_URL}: ${err}`;
    }

    recordStep("Backend health endpoint", passed, "", Date.now() - start, detail);

    if (!passed) {
      console.warn(`  [WARN] ${detail}`);
    }
  });

  // Bonus: Console Errors
  test("Bonus: No critical console errors", async () => {
    const start = Date.now();

    const criticalErrors = consoleErrors.filter((e) => {
      const text = e.text.toLowerCase();
      return (
        !text.includes("favicon") &&
        !text.includes("hmr") &&
        !text.includes("hot module") &&
        !text.includes("websocket") &&
        !text.includes("failed to load resource: net::err_connection_refused")
      );
    });

    const passed = criticalErrors.length === 0;
    const detail =
      criticalErrors.length === 0
        ? `No critical console errors (${consoleErrors.length} total, all non-critical)`
        : `${criticalErrors.length} critical console error(s) found: ${criticalErrors
            .slice(0, 3)
            .map((e) => e.text.slice(0, 100))
            .join("; ")}`;

    recordStep("No critical console errors", passed, "", Date.now() - start, detail);

    if (!passed) {
      console.warn(`  [WARN] ${detail}`);
    }
  });

  // Bonus: No API 5xx Errors
  test("Bonus: No API 5xx errors during navigation", async () => {
    const start = Date.now();

    const serverErrors = apiFailures.filter((f) => f.status >= 500);
    const passed = serverErrors.length === 0;
    const detail =
      serverErrors.length === 0
        ? `No 5xx API errors recorded (${apiFailures.length} total 4xx/5xx)`
        : `${serverErrors.length} server error(s): ${serverErrors
            .slice(0, 3)
            .map((e) => `${e.method} ${e.url} => ${e.status}`)
            .join("; ")}`;

    recordStep("No API 5xx errors", passed, "", Date.now() - start, detail);

    expect(passed, detail).toBeTruthy();
  });
});
